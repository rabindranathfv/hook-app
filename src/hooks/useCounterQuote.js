import { useState } from 'react';

export const useCounterQuote = (initualState = 1) => {
    
    const [counterQuote, setCounterQuote] = useState( initualState );

    const reset = () => {
        setCounterQuote( initualState )
    };

    const add = () => {
        setCounterQuote( counterQuote + 1 )
    };

    const subs = () => {
        setCounterQuote( counterQuote - 1 )
    }
    
    return {
        reset,
        add,
        subs,
        quoteId: counterQuote
    }
}
