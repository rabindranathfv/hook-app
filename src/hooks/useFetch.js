import { useState, useEffect, useRef } from 'react';

export const useFetch = (url) => {
    
    const isMounted = useRef(true);

    const [state, setsState] = useState({ data: null, loading: true, error: null})

    useEffect(() => {
        return () => {
            isMounted.current = false
        }
    }, [])

    useEffect(() => {
        setsState({ data: null, loading: true, error: null})

        fetch( url ).then( resp => resp.json()).then( data => {

            if (isMounted.current) {
                setsState({
                    loading: false,
                    error: null,
                    data
                })
            }
        }).catch(() => {
            setsState({
                loading: true,
                error: 'could not load phrase',
                data: null
            })
        })
    }, [url]);

    return state;
}
