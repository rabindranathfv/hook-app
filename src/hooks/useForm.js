import React, { useEffect, useState } from 'react'

export const useForm = ( initialState = { }) => {

    const [values, setValues] = useState(initialState);

    const handleInputChange = (e) => {
        setValues({
            ...values,
            [ e.target.name ] : e.target.value
        })
    }

    const reset = () => {
        setValues(initialState);
    }

    return [ values, handleInputChange, reset]
}
