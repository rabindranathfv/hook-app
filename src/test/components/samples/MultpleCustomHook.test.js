import React from 'react';
import { shallow } from 'enzyme';

import { MultpleCustomHook } from '../../../components/samples/MultpleCustomHook';
import { useFetch } from '../../../hooks/useFetch';
import { useCounterQuote } from '../../../hooks/useCounterQuote';

jest.mock('../../../hooks/useFetch',
  () => {
    return {
        useFetch: (active = true) => {
        return active ? useFetchOKMock : useFetchInitMock;
      }
    }
  }
);

const useFetchOKMock = { 
    loading: false, 
    data:  [{
        quote_id: 2,
        quote: 'Stay out of my territory.',
        author: 'Walter White',
        series: 'Breaking Bad'
    }],
    error: null
}

const useFetchInitMock = { 
    loading: true, 
    data:  null,
    error: false
}

jest.mock('../../../hooks/useCounterQuote',
  () => {
    return {
        // you can also write with return in the same sentence
        // useCounterQuote: () => ({
        //         add: () => {},
        //         quoteId: 10
        // })
        useCounterQuote: () => {
            return {
                add: () => {},
                quoteId: 10
            }
        }
    }
  }
);
describe('MultpleCustomHook::', () => {

    let wrapper = shallow( <MultpleCustomHook />);

    beforeEach(() => {
        wrapper = shallow(< MultpleCustomHook />);
    });

    afterEach(() => {
        jest.restoreAllMocks();   
    });

    it('Should render component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should should loading is false, quote OK and author OK', () => {
        useFetch();
        const loading = wrapper.find('.alert-info');
        const quote = wrapper.find('.mb-0');
        const author = wrapper.find('.blockquote-footer');
        
        expect(loading).toEqual({});
        expect(quote.text().trim()).toBe(useFetchOKMock.data[0].quote);
        expect(author.text().trim()).toBe(useFetchOKMock.data[0].author)
    });


    it('Should have add function for get next Quote', () => {
        useCounterQuote();
        const btnAdd = wrapper.find('.btn-outline-primary').at(0).simulate('click');
        expect(btnAdd.exists()).toBe(true);
    });
});