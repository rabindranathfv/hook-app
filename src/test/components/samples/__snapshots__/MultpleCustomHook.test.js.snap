// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultpleCustomHook:: Should render component 1`] = `
<div>
  <h2>
     BreakingBad Quotes
  </h2>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-0"
    >
       
      Stay out of my territory.
       
    </p>
    <footer
      className="blockquote-footer"
    >
       
      Walter White
       
    </footer>
  </blockquote>
  <button
    className="btn btn-outline-primary"
    onClick={[Function]}
    type="button"
  >
     next Quote 
  </button>
</div>
`;
